### PR TITLE
petsc: Changes for PISM

### PR DESCRIPTION
**Pull request by citibeth**

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -127,7 +134,9 @@ def install(self, spec, prefix):
                 'complex' if '+complex' in spec else 'real'),
             '--with-shared-libraries=%s' % ('1' if '+shared' in spec else '0'),
             '--with-debugging=%s' % ('1' if '+debug' in spec else '0'),
-            '--with-blas-lapack-dir=%s' % spec['lapack'].prefix
+            '--with-lapack-lib=%s' % spec['lapack'].lapack_static_lib,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: why this change? There is no static lib for MKL, so this bound to fail. I suggest to keep it as is for now.

Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: or change to `lapack_shared_lib` and `blas_shared_lib`.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -44,19 +44,26 @@ class Petsc(Package):
     version('3.5.1', 'a557e029711ebf425544e117ffa44d8f')
     version('3.4.4', '7edbc68aa6d8d6a3295dd5f6c2f6979d')
 
+    # If this has trouble finding BLAS and LAPACK...
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: i think this comment should be close to `--with-lapack-lib` so that it's easier to read and related it to the code.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 3]
```diff
@@ -127,7 +134,9 @@ def install(self, spec, prefix):
                 'complex' if '+complex' in spec else 'real'),
             '--with-shared-libraries=%s' % ('1' if '+shared' in spec else '0'),
             '--with-debugging=%s' % ('1' if '+debug' in spec else '0'),
-            '--with-blas-lapack-dir=%s' % spec['lapack'].prefix
+            '--with-lapack-lib=%s' % spec['lapack'].lapack_static_lib,
+            '--with-blas-lib=%s' % spec['blas'].blas_static_lib
+#            '--with-blas-lapack-dir=%s' % spec['lapack'].prefix
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: remove commented code.

**Comment by davydden** [comment]
nitpicking: you last commit says nothing about blas/lapack. So i would separate the changes into two 'petsc: Added consistency to code formatting` and `petsc: specify blas/lapack lib instead of a dir'

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 1 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/petsc/package.py]: @davydden Another thing that you'll see failing when using MKL is that there is no single blas or lapack library. I think we should embrace python protocols and make all singular `*_lib` plurals and iterables. For `openblas` it would be something like : 
```
self.lapack_static_libs = ('libopenblas.so', )
```
but MKL will have a long list of libraries depending on the actual case at hand. What do you think ?

Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: @alalazo you are not entirely correct, one can use perfectly fine `libmkl_rt` as a single dynamic library. As a matter of fact, as of now I can build the whole bunch of things with current MKL like Petsc, Trilinos, deal.II, mumps, etc. I have not tested them, but at least they build and link against MKL.

What you suggest is an option, but requires quite some work. Not sure we gain enough from it.

Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: @alalazo FYI the comment i recently added: https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/mkl/package.py#L49-L58

Comment (alalazo) [inline var/spack/repos/builtin/packages/petsc/package.py]: @davydden The problems I see with `libmkl_rt` are that it seems : 
- it could be driven wrong (you can select the ILP64 interface vs. LP64 whereas the size of your integers is only one in your binary files)
- it's multi-threaded by default

Anyhow this discussion may be worth a new issue on its own...

Comment (citibeth) [inline var/spack/repos/builtin/packages/petsc/package.py]: Apparently, I switched to static libs to get PISM building.  See here, it is unfortunately not very explanatory:
   https://github.com/citibeth/spack/commit/f7700e23ebedd3f01e4005ce30f3a063b3713c3d

> Better yet, change to lapack_shared_lib and blas_shared_lib if building shared petsc, otherwise use static

I submitted a PR last night for another package that does this.  Not sure why this package did not.  Unless we decide to turn those exported libs into tuples/lists, I think this is the right approach.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 2 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/petsc/package.py]: I think the comment should be removed, since we already have those smarts below.

**Comment by citibeth** [comment]
I don't know why I got two PR's in one here.  I changed the title to reflect the more serious intent of this PR.  I can separate the formatting fix into another PR, but that's also another 3 minutes PITA.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: @alalazo sure, you can break the build in run-time via wrong flags. My points was -- it is **a** solution. Whereas yours is another one.

@citibeth for the sake of this PR i would stick the shared (if defined) and static otherwise. Switching Blas/Lapack to lists will not happen immediately and requires a lot of testing.

Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: @alalazo 

> Anyhow this discussion may be worth a new issue on its own...

agree, you can create one if you wish.

**Comment by citibeth** [comment]
Looking at the `develop` branch today, I surmise that whatever issues there were with BLAS/LAPCK and shared/static, libs, a solution has been applied.  I think it's best to try that out, until/unless it is shown to cause problems.  Therefore I'm closing this PR.
